### ISSUE TITLE
undefined behaviour if multiple devices in the same group have similar capabilities (beta)

### ISSUE THREAD

**Issue by jonasBoss** [sezanzeb/input-remapper#435]
The Sony DualSense controller has a mouse-pad which reports `ABS_X` and `ABS_Y` as well as the standard controller layout which also reports `ABS_X` and `ABS_Y` events.
When a mapping uses one of those events as input both the joystick and the mouse-pad can trigger this mapping.

This leads to all sorts of issues, especially because the `absinfo` are not the same for those axis.

**Comment by jonasBoss** [comment]
https://github.com/sezanzeb/input-remapper/pull/375/commits/a622037834a824d534c8736203e3799538c9fd0b mitigates this partially by only grabbing one of the devices. 
This however does not help if both devices are grabbed by different mappings.

**Comment by jonasBoss** [comment]
@sezanzeb I think we need to include the origin of each event for a mapping. Otherwise we will not be able to fix both, this issue and #517 

You know evdev better than I do. Is there any good parameter we can use to identify a input-device, which is unique to each device in a group but stays always the same?

**Comment by sezanzeb** [comment]
I wrote some relevant stuff down at some point: https://github.com/sezanzeb/input-remapper/blob/b3e1e4ca1976bb420017c737525151ac37099e6e/inputremapper/groups.py#L219

    # - device.phys is empty sometimes and varies across virtual
    #   subdevices
    # - device.version varies across subdevices
    # - device.uniq is empty most of the time, I don't know what this is
    #   supposed to be

this is from looking for something that is constant for all 

- Maybe create a hash from all three of them
- Or sort them and use an index

**Comment by sezanzeb** [comment]
preset config

```json
{
    "1,4,1+1,6,1+1,7,1+1,8,1": {
        "target_uinput": "keyboard"
        "output_symbol": "sfdgsdfgsdfgh",
        "mapping_type": "key_macro"
    },
}
```

to

```json
[
    {
        "target_uinput": "keyboard"
        "combination": [
            {type: 1, code: 4, value: 1, device: "foo"},
            {type: 1, code: 6, value: 1, device: "foo"},
            {type: 1, code: 8, value: 1, device: "bar"}
        ],
        "output_symbol": "sfdgsdfgsdfgh",
        "mapping_type": "key_macro"
    }
]
```

?

**Comment by sezanzeb** [comment]
some more examples for the device attributes:

```
{'phys': 'usb-0000:28:00.3-2.3/input1', 'uniq': '', 'version': 65537, 'name': 'Logitech USB Keyboard System Control'}
{'phys': 'usb-0000:28:00.3-2.3/input1', 'uniq': '', 'version': 65537, 'name': 'Logitech USB Keyboard Consumer Control'}
{'phys': 'usb-0000:28:00.3-2.3/input0', 'uniq': '', 'version': 65537, 'name': 'Logitech USB Keyboard'}
```

```
{'phys': 'usb-0000:06:00.0-3/input0', 'uniq': '', 'version': 65537, 'name': 'Wacom Intuos5 M Pad'}
{'phys': 'usb-0000:06:00.0-3/input0', 'uniq': '', 'version': 65537, 'name': 'Wacom Intuos5 M Pen'}
```

In order to identify them, the name would work well for those two.

**Comment by jonasBoss** [comment]
the name will not work for a razer naga trinity, but at least the `phys` is unique
```
phys = 'usb-0000:29:00.3-1/input0', uniq = '00000000001A', version = 65537, name = 'Razer Razer Naga Trinity'
phys = 'usb-0000:29:00.3-1/input1', uniq = '00000000001A', version = 65537, name = 'Razer Razer Naga Trinity'
phys = 'usb-0000:29:00.3-1/input2', uniq = '00000000001A', version = 65537, name = 'Razer Razer Naga Trinity'
```
It honestly looks like the different devices are just a hot mess.
I am currently considering to use:
```python
hash(str(InputDevice.capabilities()))
```
given the high amount of parameters in `capabilities` I would guess that this is always unique. At least for all my devices this is always unique.

---

> ```json
>[
>    {
>        "target_uinput": "keyboard"
>        "combination": [
>            {"type": 1, "code": 4, "value": 1, "device": "foo"},
>            {"type": 1, "code": 6, "value": 1, "device": "foo"},
>            {"type": 1, "code": 7, "value": 1, "device": "foo"},
>            {"type": 1, "code": 8, "value": 1, "device": "bar"}
>        ],
>        "output_symbol": "sfdgsdfgsdfgh",
>        "mapping_type": "key_macro"
>    }
>]
>```
yeah, I was thinking something like this as well.

**Comment by sezanzeb** [comment]
yeah. Since this only happens once when the injection starts, the complexity of the hash generation doesn't matter.

```py
hash(str(InputDevice.capabilities()) + InputDevice.phys + InputDevice.name)
```
